### ISSUE TITLE
screen_width() and screen_height() return window size, but not fullscreen size

### ISSUE THREAD

**Issue by nacl42** [not-fl3/macroquad#237]
When in fullscreen mode, screen_width() and screen_height() return the size of the window, but not the size of the full screen. It would be useful to return the actual available space for drawing.

**Comment by not-fl3** [comment]
hmm, this may be a very serious bug, thanks for reporting!

Unfortunately, I can't really reproduce it:

```rust
use macroquad::prelude::*;

fn window_conf() -> Conf {
    Conf {
        window_title: "Window Conf".to_owned(),
        fullscreen: true,
        ..Default::default()
    }
}

#[macroquad::main(window_conf)]
async fn main() {
    loop {
        clear_background(BLACK);
        draw_rectangle(5., 5., screen_width() - 10., screen_height() - 10., RED);
        next_frame().await
    }
}
```

![image](https://user-images.githubusercontent.com/910977/120905333-fbbd4f00-c616-11eb-91c5-e2348f011944.png)


This rectangle takes all the space available :( 

Could you give a reproducible example and some info about the platform you are on?

**Comment by nacl42** [comment]
Thanks for your quick response.

Actually, my example is rather complicated (see nacl42/reveal).

It seems, that when I start my example several times, then two out of three times, the full screen size is recognized, while for one out of three times, the smaller window size is returned. Since you reported that this shouldn't occur, I switched my window manager from XFCE to GNOME, and surprisingly everything works fine there. So this seems to be a problem with xfwm4 and not necessarily with macroquad.

**Comment by martin-t** [comment]
I can confirm this happens randomly on Kubuntu 20.10 - maybe one time out of 20. A quicker way to repro is something like this:
```rust
use macroquad::prelude::*;

fn window_conf() -> Conf {
    Conf {
        window_title: "Window Conf".to_owned(),
        fullscreen: true,
        ..Default::default()
    }
}

#[macroquad::main(window_conf)]
async fn main() {
    dbg!(screen_width());
    assert_eq!(screen_width(), 1920.0);
}
```
Run it multiple times until you get a failure.

**Comment by not-fl3** [comment]
> I can confirm this happens randomly on Kubuntu 20.10 - maybe one time out of 20. A quicker way to repro is something like this:
> 
> ```rust
> use macroquad::prelude::*;
> 
> fn window_conf() -> Conf {
>     Conf {
>         window_title: "Window Conf".to_owned(),
>         fullscreen: true,
>         ..Default::default()
>     }
> }
> 
> #[macroquad::main(window_conf)]
> async fn main() {
>     dbg!(screen_width());
>     assert_eq!(screen_width(), 1920.0);
> }
> ```
> 
> Run it multiple times until you get a failure.

is it the same value after a few frames? Maybe window manager is doing weird things while stretching the window?

**Comment by martin-t** [comment]
Yeah, seems like after a few frames it starts reporting the correct fullscreen size. It still complicates things for me though because I need to create render targets with the correct size during init so I'd need to resize them later.

**Comment by martin-t** [comment]
Since this probably can't be fixed cleanly, would it make sense to use some kind of ugly workaround? I am thinking macroquad could wait a few frames and check the window size before giving control to user code (the contents of `main`).

- It would only do that when fullscreen is requested so there'd be no additional startup time for non-fullscreen games.
- In my tests, when the size is wrong, it always corrects itself on the 3rd frame (interestingly, never earlier - it's either correct from the start or 3rd frame) so there could be a max timeout of say 100 ms to be sure.
- It could use some magic value for `window_width` and `window_height` that is sure not to be fullscreen on any sane device (e.g. 1x100 px), then it could check what `screen_width()` and `screen_height()` return so that it could give control to user code even before the timeout elapses if startup time is a concern.

**Comment by emko** [comment]
anyone have a workaround? i need to create a render target but the screen_width and height are wrong

**Comment by tjpalmer** [comment]
> anyone have a workaround? i need to create a render target but the screen_width and height are wrong

[Here's my burn frames loop](https://github.com/contextfreeinfo/sing/blob/c05e9e9e57997f29b6df290ced8016ba254dadf5/src/main.rs#L67) based on the recommendation of martin-t above.

Note that I don't actually do any app logic at all until after a few frames.

**Comment by emko** [comment]
> > anyone have a workaround? i need to create a render target but the screen_width and height are wrong
> 
> [Here's my burn frames loop](https://github.com/contextfreeinfo/sing/blob/c05e9e9e57997f29b6df290ced8016ba254dadf5/src/main.rs#L67) based on the recommendation of martin-t above.
> 
> Note that I don't actually do any app logic at all until after a few frames.

what i did was save the current wdith,height into mutable variables and then in my loop i am comparing those to the current and if they are different i update them and update any render targets to the new value, this helps also with when the app is not in fullscreen and user resizes the window.
